addons/costume-editor-shape-ops/userscript.js

```diff
@@ -0,0 +1,452 @@
+    }
+    return null;
+  };
+  // https://github.com/scratchfoundation/scratch-paint/blob/develop/src/helper/compound-path.js
+  const isCompoundPath = function (item) {
+    return item && item.className === "CompoundPath";
+  };
+  const getItemsCompoundPath = function (item) {
+    const itemParent = item.parent;
+
+    if (isCompoundPath(itemParent)) {
+      return itemParent;
+    }
+    return null;
+  };
+  const _setGroupSelection = function (root, selected, fullySelected) {
+    root.fullySelected = fullySelected;
+    root.selected = selected;
+    // select children of compound-path or group
+    if (isCompoundPath(root) || isGroup(root)) {
+      const children = root.children;
+      if (children) {
+        for (const child of children) {
+          if (isGroup(child)) {
+            _setGroupSelection(child, selected, fullySelected);
+          } else {
+            child.fullySelected = fullySelected;
+            child.selected = selected;
+          }
+        }
+      }
+    }
+  };
+
+  const setItemSelection = function (item, state, fullySelected) {
+    const parentGroup = getItemsGroup(item);
+    const itemsCompoundPath = getItemsCompoundPath(item);
+
+    // if selection is in a group, select group
+    if (parentGroup) {
+      // do it recursive
+      setItemSelection(parentGroup, state, fullySelected);
+    } else if (itemsCompoundPath) {
+      _setGroupSelection(itemsCompoundPath, state, fullySelected);
+    } else {
+      if (item.data && item.data.noSelect) {
+        return;
+      }
+      _setGroupSelection(item, state, fullySelected);
+    }
+  };
+
+  function decomposeCompoundPath(path) {
+    const childIndex = path.parent.children.indexOf(path);
+    const items = path.parent.insertChildren(childIndex, path.children);
+    path.remove();
+    return items || [];
+  }
+  function recursiveDecompose(item, dryRun = false) {
+    if (Array.isArray(item)) {
+      const newArray = [];
+      for (const child of item) {
+        newArray.push(...recursiveDecompose(child, dryRun));
+      }
+      return newArray;
+    }
+    if (!item || !item.children) return [item];
+    const children = item.children;
+    if (!dryRun) decomposeCompoundPath(item);
+    return children.map(() => recursiveDecompose(child, dryRun));
+  }
+
+  // Some of this code was originally written by JeremyGamer13 for PenguinMod,
+  // and further modified by CST1229
+  // https://github.com/CST1229/scratch-paint/blob/0283b94a479d58af9500af0317f6f1bc0f193f78/src/containers/mode-tools.jsx#L191-L347
+  function handleMergeShape(specificOperation, doSelections) {
+    // if it breaks, it's probably better to throw a debuggable error in the console rather than silently failing
+    const modeToolsEl = document.querySelector("[class*='paint-editor_mod-mode-tools_']");
+    const internalInstanceKey = addon.tab.traps.getInternalKey(modeToolsEl);
+    const modeTools = modeToolsEl[internalInstanceKey]?.child?.child?.child?.child?.stateNode;
```

Comment:
I haven't tested this with React 18, but it will probably break. Instead of reading `.child` a fixed number of times, it's better to use a loop to traverse the component tree until a component with a specific prop or method is found. Example code in another addon: https://github.com/mxmou/ScratchAddons/blob/a9a8c4b/addons/onion-skinning/userscript.js#L14-L18